### PR TITLE
Handle errors on finalizing draft order

### PR DESCRIPTION
**Pull request by orzechdev**
It adds errors handling in the draft order view.

**PR intended to be tested with API branch:**  main

### Screenshots

Screenshot 2022-07-25 at 16 19 50

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://marketplace-gray.vercel.app/

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-1879 Handle Errors on finalizing to Draft Order](https://app.clickup.com/t/2549495/SALEOR-1879)

File: src/orders/components/OrderDraftDetailsProducts/OrderDraftDetailsProducts.tsx [hunk 1]
```diff
@@ -20,60 +27,65 @@ export interface FormData {
...
 
 const OrderDraftDetailsProducts: React.FC<OrderDraftDetailsProductsProps> = props => {
-  const { lines, onOrderLineChange, onOrderLineRemove } = props;
+  const { order, errors, onOrderLineChange, onOrderLineRemove } = props;
+  const lines = order?.lines ?? [];
 
+  const intl = useIntl();
   const classes = useStyles(props);
 
+  const formErrors = errors.filter(error => error.field === "lines"); // getFormErrors(["lines"], errors);
```
Comment (Cloud11PL) [inline src/orders/components/OrderDraftDetailsProducts/OrderDraftDetailsProducts.tsx:77]: The comment is not necessary I think

File: src/orders/components/OrderDraftDetailsProducts/TableLineAlert.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    return null;
+  }
+
+  const title =
+    alerts.length === 1 ? (
+      <>{alerts[0]}</>
+    ) : (
+      <ul>
+        {alerts.map((alert, index) => (
+          <li key={index}>{alert}</li>
```
Comment (Cloud11PL) [inline src/orders/components/OrderDraftDetailsProducts/TableLineAlert.tsx:24]: Maybe like this?
```suggestion
          <li key={`${index}_${alert}`}>{alert}</li>
```

File: src/orders/components/OrderDraftDetailsProducts/messages.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  notPublished: {
+    id: "Oad+ES",
+    defaultMessage: "This product is not published in this channel.",
+    description: "alert message",
+  },
+  notAvailable: {
+    id: "zO+l0L",
+    defaultMessage: "This product is not available for sale in this channel.",
+    description: "alert message",
+  },
```
Comment (Cloud11PL) [inline src/orders/components/OrderDraftDetailsProducts/messages.ts:13]: The same message is in `locale/defaultMessages.json`

File: src/orders/components/OrderDraftPage/OrderDraftAlert.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { FormattedMessage, MessageDescriptor } from "react-intl";
+
+import { alertMessages } from "./messages";
+import { useAlertStyles } from "./styles";
+
+const getAlerts = (
+  order?: OrderDetailsFragment,
+  channelUsabilityData?: ChannelUsabilityDataQuery,
+) => {
+  const inactiveChannel = order && !order.channel.isActive;
```
Comment (Cloud11PL) [inline src/orders/components/OrderDraftPage/OrderDraftAlert.tsx:16]:
```suggestion
  const isChannelInactive = order && !order.channel.isActive;
```
or just `isChannelActive`

File: src/orders/components/OrderDraftPage/OrderDraftAlert.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            {alerts.map((alert, index) => (
+              <li key={index}>
+                <FormattedMessage {...alert} />
+              </li>
+            ))}
+          </ul>
+        </>
+      ) : (
+        <FormattedMessage {...alerts[0]} />
+      )}
```
Comment (Cloud11PL) [inline src/orders/components/OrderDraftPage/OrderDraftAlert.tsx:65]: It's very similar logic to what's in `src/orders/components/OrderDraftDetailsProducts/TableLineAlert.tsx`. Maybe we can unify it somehow (if it makes sense ofc)

**Review by Cloud11PL — APPROVED** [review]

File: src/orders/components/OrderDraftDetailsProducts/messages.ts (hunk 3 above)
Comment (orzechdev) [reply src/orders/components/OrderDraftDetailsProducts/messages.ts:13]: `locale/defaultMessages.json` are auto-generated on purpose for service like Transifex from all ts/tsx files.

Comment (Cloud11PL) [reply src/orders/components/OrderDraftDetailsProducts/messages.ts:13]: Ah, alright!

File: src/orders/components/OrderDraftPage/OrderDraftAlert.tsx (hunk 5 above)
Comment (orzechdev) [reply src/orders/components/OrderDraftPage/OrderDraftAlert.tsx]: Ok, I refactored it a bit

Comment (Cloud11PL) [reply src/orders/components/OrderDraftPage/OrderDraftAlert.tsx]: Looks good 👍

File: src/orders/components/OrderCustomer/AddressFields.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Typography } from "@material-ui/core";
```
Comment (dominik-zeglen) [inline src/orders/components/OrderCustomer/AddressFields.tsx:1]: I thought we have address form component somewhere

File: src/orders/components/OrderCustomer/OrderCustomer.tsx [hunk 7]
```diff
@@ -121,6 +110,15 @@ const OrderCustomer: React.FC<OrderCustomerProps> = props => {
...
   };
 
+  const addressTextError = (orderError: OrderErrorFragment) => (
+    <>
+      <Typography variant="body2" className={classes.textError}>
+        {getOrderErrorMessage(orderError, intl)}
+      </Typography>
+      <FormSpacer />
+    </>
+  );
```
Comment (dominik-zeglen) [inline src/orders/components/OrderCustomer/OrderCustomer.tsx:120]: Either make it component or copy it, component factories are antipattern in react

File: src/orders/components/OrderDraftDetailsProducts/useLineAlerts.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    alerts.push(getOrderErrorMessage(error, intl));
+  }
+  if (!isPublished) {
+    alerts.push(intl.formatMessage(lineAlertMessages.notPublished));
+  }
+  if (!isAvailable) {
+    alerts.push(intl.formatMessage(lineAlertMessages.notAvailable));
+  }
+
+  return alerts;
```
Comment (dominik-zeglen) [inline src/orders/components/OrderDraftDetailsProducts/useLineAlerts.ts:39]: If you're making it a hook then please at least memoize its output

**Review by dominik-zeglen — CHANGES_REQUESTED** [review]

**Comment by timuric** [comment]
Amazing PR!

Some edge cases:

I get this message 
image

but then it disappears when I add customer address
image

**Comment by timuric** [comment]
Maybe we should replace error message with something more friendly. 

`No products are available in channel of this order`
`Channel_name`

image

**Comment by orzechdev** [comment]
@timuric 
Ad.1. Changed. The warning is now only returned when shipping address and at least one line is set, as backend cannot determine what shipping methods are available without these data. This is why the warning could be wrongly displayed.
Ad.2. Changed appropriately:
Screenshot 2022-08-04 at 14 04 32
Screenshot 2022-08-04 at 14 04 46

**Comment by patrys** [comment]
Could we explicitly state the channel when we're signaling a potentially broken setup? If my colleague sent me this screenshot, there is no way for me to guess what needs to be fixed.

**Comment by orzechdev** [comment]
I've updated the messages:
Screenshot 2022-08-05 at 14 26 51
Screenshot 2022-08-05 at 14 27 15

Also I've also changed channel name from text to link and move this section to the top ("Channel without products" is a channel name here):
Screenshot 2022-08-05 at 14 45 40

I would stay with messages "... channel assigned to this order" instead of explicitly naming channel to indicate what this channel has to do with this order.

Is it better right now @patrys ?

**Comment by patrys** [comment]
I think it's good.

**Review by dominik-zeglen — APPROVED** [review]

**Pull request merged**